helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -183,19 +185,29 @@ private double getAssignmentNormalizedScore(AssignableNode node, AssignableRepli
       computeScore(overallClusterRemainingCapacityMap);
     }
 
-    public void computeScore(Map<String, Integer> overallClusterRemainingCapMap) {
+    public void computeScore(Map<String, Integer> overallClusterRemainingCapMap)
+        throws HelixRebalanceException {
       float score = 0;
       // score = SUM(weight * (resource_capacity/cluster_capacity) where weight = 1/(1-total_util%)
       // it could be be simplified to "resource_capacity/cluster_remainingCapacity".
       for (Map.Entry<String, Integer> resourceCapacity : _replica.getCapacity().entrySet()) {
+        String capacityKey = resourceCapacity.getKey();
         if (resourceCapacity.getValue() == 0) {
           continue;
         }
-        score = (overallClusterRemainingCapMap.get(resourceCapacity.getKey()) == 0
+        if (overallClusterRemainingCapMap.get(capacityKey) < 0) {
+          // all replicas' assignment will fail if there is one dimension's remain capacity <0.
+          String errorMessage = String.format(
+              "The cluster does not have enough %s capacity for all partitions. ",
+              capacityKey);
+          throw new HelixRebalanceException(errorMessage,
+              HelixRebalanceException.Type.FAILED_TO_CALCULATE);
+        }
+        score = (overallClusterRemainingCapMap.get(capacityKey) == 0
```

Comment:
Can this be moved out from `AssignableReplicaWithScore.computeScore`? 
It seems it can be checked before any score is computed, like a validation step. For one thing, I feel we are doing a lot of duplicated check on the same `capacityKey`. Besides, we don't need to compute any score if the assignment will fail.

Reply:
Do you mean the earlier check for conditional Float.MAX_VALUE score? 

I think this check is not necessary given the earlier if condition on line 198. @xyuanlu

Reply:
@NealSun96  not really, I mean `overallClusterRemainingCapMap.get(capacityKey) < 0` may be checked outside of `AssignableReplicaWithScore` and decoupled with computeScore. We should be able to construct an aggregated `capacityKey` for all replicas and check against them.

Reply:
Oh I see what you mean @qqu0127 . I agree that the check on line 198 may be checked outside. 

I think my second point still stands @xyuanlu , the check on line 206 may not be necessary.

Reply:
TFTR. Good suggestion.